mos-platform/cx16/cx16.h

```diff
@@ -118,29 +120,69 @@ extern "C" {
 #define COLOR_LIGHTBLUE         0x0E
 #define COLOR_GRAY3             0x0F
 
-/* NES controller masks for joy_read() */
-
-#define JOY_BTN_1_MASK          0x80
-#define JOY_BTN_2_MASK          0x40
-#define JOY_BTN_3_MASK          0x20
-#define JOY_BTN_4_MASK          0x10
-#define JOY_UP_MASK             0x08
-#define JOY_DOWN_MASK           0x04
-#define JOY_LEFT_MASK           0x02
-#define JOY_RIGHT_MASK          0x01
+/// NES controller masks for JoyState
+enum : unsigned char {
+  // Masks for JoyState::data0
+  JOY_BTN_B_MASK = 1 << 7,
+  JOY_BTN_Y_MASK = 1 << 6,
+  JOY_SELECT_MASK = 1 << 5,
+  JOY_START_MASK = 1 << 4,
+  JOY_UP_MASK = 1 << 3,
+  JOY_DOWN_MASK = 1 << 2,
+  JOY_LEFT_MASK = 1 << 1,
+  JOY_RIGHT_MASK = 1 << 0,
+  // Masks for JoyState::data1
+  JOY_BTN_A_MASK = 1 << 7,
+  JOY_BTN_X_MASK = 1 << 6,
+  JOY_FIRE_LEFT_MASK = 1 << 5,
+  JOY_FIRE_RIGHT_MASK = 1 << 4
+};
 
-#define JOY_BTN_A_MASK          JOY_BTN_1_MASK
-#define JOY_BTN_B_MASK          JOY_BTN_2_MASK
-#define JOY_SELECT_MASK         JOY_BTN_3_MASK
-#define JOY_START_MASK          JOY_BTN_4_MASK
+/// Joystick selection
+#define JOY_KEYBOARD 0
+#define JOY_SNES_PORT1 1
+#define JOY_SNES_PORT2 2
+#define JOY_SNES_PORT3 3
+#define JOY_SNES_PORT4 4
 
-#define JOY_BTN_A(v)            ((v) & JOY_BTN_A_MASK)
-#define JOY_BTN_B(v)            ((v) & JOY_BTN_B_MASK)
-#define JOY_SELECT(v)           ((v) & JOY_SELECT_MASK)
-#define JOY_START(v)            ((v) & JOY_START_MASK)
+/// Status of SNES joystick populated by cx16_k_joystick_get();
+struct JoyState {
+  union {
+    struct {
+      unsigned char data0; //!< Bits: B Y Select Start Up Down Left Right
+      unsigned char data1; //!< Bits: A X FireL  FireR 1  1    1    1
+    };
+    unsigned short data; //!< data0 and data1 combined
+  };
+  bool detached; //!< True if joystick is disconnected
 
-#define JOY_FIRE2_MASK          JOY_BTN_2_MASK
-#define JOY_FIRE2(v)            ((v) & JOY_FIRE2_MASK)
+#ifdef __cplusplus
+  inline bool button_a() const { return !(data1 & JOY_BTN_A_MASK); }
```

Comment:
`inline` is superfluous: all definitions directly in structs/classes are `inline`

Reply:
Removed inline keywords.